presto-hive/src/main/java/io/prestosql/plugin/hive/HiveQlToPrestoTranslator.java

```diff
@@ -0,0 +1,168 @@
+        // their delimiters are encountered.
+        while (input.hasNext()) {
+            char c = input.next();
+            switch (c) {
+                case '"':
+                case '\'':
+                    translateString(c);
+                    break;
+                case '`':
+                    translateQuotedIdentifier();
+                    break;
+                default:
+                    output.append(c);
+                    break;
+            }
+        }
+
+        return output.toString();
+    }
+
+    /**
+     * Start with input just after the opening backquote, consume to the
+     * closing backquote.
+     */
+    private void translateQuotedIdentifier()
+    {
+        output.append('"');
+        while (input.hasNext()) {
+            char c = input.next();
+
+            // The possible outcomes in an identifier are:
+            // - replace " with ""
+            // - leave most characters as-is
+            // - replace `` with `
+            // - replace ` with " and end the identifier
+            if (c == '"') {
+                output.append("\"\"");
+            }
+            else if (c != '`') {
+                output.append(c);
+            }
+            else if (input.hasNext() && peek() == '`') {
+                output.append('`');
+                input.next();
+            }
+            else { // end of identifier
+                output.append('"');
+                return;
+            }
+        }
+
+        throw hiveParseError("unexpected end of input in identifier");
+    }
+
+    /**
+     * Call with input just after the opening delimiter (the parameter), end
+     * with input after closing delimiter.
+     */
+    private void translateString(char delimiter)
+    {
+        output.append("'");
+        while (input.hasNext()) {
+            char c = input.next();
+
+            // The possible outcomes in a string are (assuming "-quoting):
+            // - find delimiter and end the string
+            // - find backslash and treat following character as non-delimiter
+            // - replace ' with ''
+            // - leave any other character as-is
+            if (c == delimiter) {
+                output.append("'");
+                return;
+            }
+            else if (c == '\\') {
+                if (!input.hasNext()) {
+                    break; // skip to end-of-input error
+                }
+                // Don't handle escape sequences, just drop the backslash.
+                // (If we want to handle them, make this block a function.)
+                c = input.next();
+            }
+
+            if (c == '\'') {
+                output.append("''");
+            }
+            else {
+                output.append(c);
+            }
+        }
+        throw hiveParseError("unexpected end of input in string");
+    }
+
+    /** Get the next character in the input without consuming it. */
+    private char peek()
```

Comment:
Use `PeekingIterator` instead? Then you don't have to store `source`

Reply:
I spent way too long trying to find a better way to handle the input, and I never found that class. Thanks!